ocaml/xapi/eventgen.ml

```diff
@@ -11,248 +11,200 @@
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU Lesser General Public License for more details.
  *)
-module D = Debug.Make (struct let name = "sql" end)
+open Debug.Make (struct let name = "sql" end)
 
-open D
-
-type getrecord = unit -> Rpc.t
+type get_record = unit -> Rpc.t
 
 let get_record_table :
-    (string, __context:Context.t -> self:string -> getrecord) Hashtbl.t =
-  Hashtbl.create 20
+    (string, __context:Context.t -> self:string -> get_record) Hashtbl.t =
+  Hashtbl.create 64
+
+let set_get_record = Hashtbl.replace get_record_table
 
-let find_get_record x ~__context ~self () : Rpc.t option =
+let find_get_record obj_name ~__context ~self () : Rpc.t option =
   Option.map
-    (fun x -> x ~__context ~self ())
-    (Hashtbl.find_opt get_record_table x)
+    (fun f -> f ~__context ~self ())
+    (Hashtbl.find_opt get_record_table obj_name)
 
-(* If a record is created or destroyed, then
-   for any (Ref _) field which is one end of a relationship, need to send
-   modified events for all those other objects. *)
-(* we build a hashtable of these references and then look them up by object on each db write: *)
+(* If a record is modified, events must be emitted for related objects' records.
+   We collect a list of related objects by querying the (Ref _)-typed
+   fields of the input object against the relations encoded by the datamodel.
+
+   The result of this function is a list of pairs [(object, field);, ...].
+   Note that the field component refers to a field in the input object,
+   not the related object. *)
 let compute_object_references_to_follow (obj_name : string) =
+  let module DT = Datamodel_types in
   let api = Datamodel.all_api in
-  let objs = Dm_api.objects_of_api api in
-  let obj = List.find (fun obj -> obj.Datamodel_types.name = obj_name) objs in
-  let relations = Dm_api.relations_of_api api in
-  let symmetric = List.concat_map (fun (a, b) -> [(a, b); (b, a)]) relations in
-  let set = Xapi_stdext_std.Listext.List.setify symmetric in
-  List.concat_map
-    (function
-      | {
-          Datamodel_types.ty= Datamodel_types.Ref _
-        ; Datamodel_types.field_name
-        ; _
-        } ->
-          let this_end = (obj.Datamodel_types.name, field_name) in
-          if List.mem_assoc this_end set then
-            let other_end = List.assoc this_end set in
-            let other_obj = fst other_end in
-            [(other_obj, field_name)]
-          else
-            []
-      | _ ->
-          []
-      )
-    (Datamodel_utils.fields_of_obj obj)
+  let obj = Dm_api.get_obj_by_name api ~objname:obj_name in
```

Comment:
I assume any such relations are small such that lists don't create an inefficiency.

Reply:
If this cost is incurred once, I don't care about it. If you suspect it takes long, maybe emit time stamps before/after or calculate the duration and log it.